00X-tracing.md

```diff
@@ -0,0 +1,581 @@
+   *
+   * @param category Required tracing category name or array of one or more category names for
+   *   the tracing event.
+   * @param name Required name for the event.
+   * @param id Required integer scope identifier that can be used to correlate paired "begin"
+   *   and "end" events and distinguish them from other events with the same name. This must be
+   *   the same as a scope id that was generated by a call to the "begin" method. This scope id
+   *   may be null if the "begin" call did not emit an event; in that case this call is also a
+   *   no-op.
+   * @param args Optional dictionary of name-value pairs, or a function that returns a dictionary.
+   *   If a function, it will be invoked to retrieve the args only if tracing is enabled for the
+   *   category.
+   * @returns True if the event was emitted; false if the event was not emitted because tracing
+   *   was not enabled for any of the event categories.
+   */
+  end(
+    category: string | string[],
+    name: string,
+    id: number | null,
+    args?: { [name: string]: any } | () => { [name: string]: any }): boolean;
+
+  /**
+   * Emits a tracing event of type "count" using an event name and optional id and args.
+   *
+   * @param category Required tracing category name or array of one or more category names for
+   *   the tracing event.
+   * @param name Required name for the event.
+   * @param value Required counter value, either a number, a dictionary of name-number pairs, or a
+   * function that returns one of those. If a function, it will be invoked to retrieve the value(s)
+   * only if tracing is enabled for the category. (The tracing system currently requires
+   * multi-value counters to have exactly two values.)
+   * @returns True if the event was emitted; false if the event was not emitted because tracing
+   *   was not enabled for any of the event categories.
+   */
+  count(
+    category: string | string[],
+    name: string,
+    value: number | { [name: string]: number } | () => (number | { [name: string]: number })): boolean;
+}
+```
+
+## 3. Tracing listener APIs
+The tracing listener APIs enable JavaScript code to listen to near-real-time tracing events in
+specific categories, rather than waiting for the events be recorded to a file for offline analysis.
+Tracing events emitted in both JavaScript can C++ can be received by both JavaScript and C++
+listeners.
+
+Relatively few applications and libraries are expected to use the listener APIs (as compared
+to the much broader applicability of the emitter APIs). However, the ability to listen to tracing
+events could be very valuable to certain application-monitoring, profiling, and diagnostic tools
+and SDKs.
+
+To support listeners, the `Tracing` class (from the previous section) extends a new
+`CategoryEventEmitter` base class. That class is similar to `EventEmitter`, but does not extend
+`EventEmitter` because it has different behavior related to the multi-category support.
```

Comment:
I considered trying to make it more different from `EventEmitter`, but it felt silly to use different names for operations that are really the same as the normal `EventEmitter` methods.

I think the differences probably should not really be a problem. You can in fact use a `CategoryEventEmitter` (which the `Tracing` module extends) just like a normal `EventEmitter` if you don't care about the multi-category support.

Would it help to think of `CategoryEventEmitter` as a *subclass* of `EventEmitter`? (I considered literally extending the class, but it didn't provide any benefit since all the methods needed to be overridden to use different data structures from the base class.)